### PR TITLE
Changes to Synchrotron Radiation from issue #4921. Part 1.

### PR DESCRIPTION
**Pull request by FilipO28555**
in validate():
- [x] subtract_functions() is now without a bug and unnecessary list truncation
- [x] Changed the number of bins to 200 for better accuracy

Todos from [issue #4912](https://github.com/ComputationalRadiationPhysics/picongpu/issues/4912):
- [x] use `pathlib` instead of `os.path` [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609863570) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609864862)
- [x] [question on usage](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609879390) of `change_parameters_randomly.py`
- [x] use [`main` function](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609881468) instead of a script like code
- [x] [question on comment](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609882515) regarding output 
- [x] use of `np.asarray` [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609891973) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609899282)
- [x] [question on parameters](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609902008) in `validate.py`
- [x] use seperate functions in python [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609908954) and [here](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609909367)
- [x] better way to handle [input arguments](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4827#discussion_r1609940439)

### CODE REVIEW

File: share/picongpu/tests/Synchrotron_plugin/lib/change_parameters_randomly.py [hunk 1]
```diff
@@ -23,6 +24,7 @@ def change_param_file(file_name, param_name, value):
                 f.write(line)
 
 
+# change the three parameters
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/change_parameters_randomly.py]: Remove. That is redundant.

File: share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt [hunk 2]
```diff
@@ -1,4 +1,5 @@
 os
-numpy
-scipy
-openpmd_api
+numpy==1.26.4
+scipy==1.11.4
+pathlib
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt]: `pathlib` is part of the standard library, so we don't need to include it.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py [hunk 3]
```diff
@@ -7,15 +7,8 @@
...
 def subtract_functions(x0, y0, x1, y1):
-    minX = max(np.min(x0), np.min(x1))
-    maxX = min(np.max(x0), np.max(x1))
-    #
-    x0 = [x for x in x0 if x >= minX and x <= maxX]
-    x1 = [x for x in x1 if x >= minX and x <= maxX]
-    y0 = [y for x, y in zip(x0, y0) if x >= minX and x <= maxX]
-    y1 = [y for x, y in zip(x1, y1) if x >= minX and x <= maxX]
-
     y1 = np.interp(x0, x1, y1)  # interpolate y1 to the x0 values
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: Okay, so you are saying that you are fine with the default behaviour of `np.interp` here? I would have assumed that for histograms using `left=0, right=0` would have been the most faithful extrapolation.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py [hunk 4]
```diff
@@ -38,57 +31,84 @@ def read_setup(folderName):
...
     series.flush()
     p_y = np.float64(p_y)
     p_y *= p_y_SI / w
-    hist_data = np.abs(p_y)
+
+    # Calculate photon energy
+    hist_data = np.abs(p_y * const.c / const.elementary_charge)
+
     if len(hist_data) < 5e5:
         raise SystemExit(
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: Just noticed: Do we really want to raise `SystemExit` here? Raising something a little milder like `ValueError` or a custom `Exception` would have the same effect for running this script. But in case anyone ever tries to build something on top of this, a `ValueError` could be caught and handled appropriately.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py [hunk 5]
```diff
@@ -38,57 +31,84 @@ def read_setup(folderName):
...
+
+    # Reading electron weights
     h1 = e["weighting"][io.Mesh_Record_Component.SCALAR]
     e_w = h1[:]  # get the weighting
     e_w_SI = h1.unit_SI  # get the weighting unit
     series.flush()
     e_w *= e_w_SI  # convert to SI units
+
+    # Reading electron momentum
     h1 = e["momentum"]["y"]
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/validate.py:76]: Just to be extra sure: The `"y"` in this line has nothing to do with the `"y"` that approximates a $\gamma$ and stands for photons further up, right? It's probably the direction of the momentum?

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py [hunk 6]
```diff
@@ -114,11 +134,12 @@ def main(dataPath):
     analytical_integrated = np.array(analytical_integrated)[mask]
 
     x, y = subtract_functions(delta, analytical_integrated, b, a)
-    poorness = np.sum(y) / len(y)  # average error
+    poorness = np.sum(y) / len(y)  # average relative error
```
Comment (chillenzer) [inline share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: Use `np.mean` (potentially with `axis=0`) here.

**Review by chillenzer — COMMENTED** [review]
Hi, looks good. Thanks for going the extra mile to make this a little more sustainable code. In what sense are the questions answered? They are already ticked but that information has not propagated to me. So, it is likely that whatever method you used to answer them, does not help the uninvolved future reader of your code.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 5 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py:76]: yes, you are right

File: share/picongpu/tests/Synchrotron_plugin/lib/change_parameters_randomly.py (hunk 1 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/change_parameters_randomly.py]: done

File: share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt (hunk 2 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt]: Done

Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt]: The same with os, right?

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 3 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: You are right. I did that. I also changed the name to be more descriptive

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 4 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: I changed it to ValueError

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 5 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py:76]: Yes

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 6 above)
Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: We left it because in the future we might do something like np.sum(y*np.sqrt(a)) / len(y)

File: share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt (hunk 2 above)
Comment (chillenzer) [reply share/picongpu/tests/Synchrotron_plugin/lib/requirements.txt]: Yes

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 6 above)
Comment (chillenzer) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: That's still a mean isn't it? `np.mean(y*np.sqrt(a))`

**Review by chillenzer — APPROVED** [review]
Good work!

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 6 above)
Comment (PrometheusPi) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: I agree with @chillenzer

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py [hunk 7]
```diff
@@ -113,17 +133,18 @@ def main(dataPath):
...
     retValue = int(0) if poorness <= poornessBound else int(1)
+    print(f"Test {'passed' if retValue == 0 else 'failed'}")
 
     sys.exit(retValue)
 
 
 if __name__ == "__main__":
     if len(sys.argv[1:]) != 1:
-        raise SystemExit(f"Usage: {sys.argv[0]} <path_to_simulation_data>")
+        raise ValueError(f"Usage: {sys.argv[0]} <path_to_simulation_data>")
```
Comment (PrometheusPi) [inline share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: Just a question: why this change?

**Review by PrometheusPi — CHANGES_REQUESTED** [review]
Just some minor requests.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 7 above)
Comment (chillenzer) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: I guess that was in response to [this comment](https://github.com/ComputationalRadiationPhysics/picongpu/pull/4926#discussion_r1629120358) applying to a different exception in the same script. The change you are referring to will most likely not make any difference at all and one could argue that not doing it might convey the intent a tiny bit better. So, weakly agreeing with you but probably doesn't matter.

File: share/picongpu/tests/Synchrotron_plugin/lib/validate.py (hunk 6 above)
Comment (PrometheusPi) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: and why is this now "relative" ?

Comment (FilipO28555) [reply share/picongpu/tests/Synchrotron_plugin/lib/validate.py]: It was always relative.
Also changed to np.mean()

**Review by PrometheusPi — APPROVED** [review]
